### ISSUE TITLE
IO module: how should we adjust openfp?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#20146]
It has the signature:
```chapel
 proc openfp(fp: _file, hints: iohints = IOHINT_NONE): file throws
```
And creates a `file` that points to the C file pointer provided by the `fp` argument.

Note that `_file` is a primitive type inserted into the program by the compiler, and represents the C type `FILE *`.  It seems like it would be more appropriate for this type to be an extern type defined in CTypes, if that's something we can do.  It probably should also be named `c_FILE` or `C_FILE` or `c_file` in that case.

We think it would be reasonable to keep this function.  If `_file` does get moved to CTypes, should this function also be moved there?

It also seems like this function should be renamed.  Similarly to the options proposed by #20143, here are some ideas we came up with:
- openFP
- openFp
- openFilePtr
- openFilePointer
- make it a factory function on the `file` type, e.g. `file.fromFP` or `file.fromFilePtr`, since it's already open as a result of how it is obtained?
- make it an initializer on the `file` type, for a similar reason?

This function also notes in its documentation that the I/O system will ignore the channel offsets when reading or writing to a file opened by this function, so it would be good to have some way for the `file` object to know that and generate a warning when someone tries to open a channel with an offset

This function throws a subset of the errors thrown by `open` (as discussed in #20139).  Similar updates should be performed to its documentation.

Note that #20430 will also impact this function

**Comment by lydia-duncan** [comment]
Should we make `_file` be an extern type defined in CTypes, if possible?  Vote here (👍🏻 👎🏻 )

**Comment by lydia-duncan** [comment]
Vote here for what to rename `openfp` to:
👍🏻 - openFP
😄 - openFp
🎉 - openFilePtr
❤️ - make it a factory function on the `file` type (will result in a separate discussion about the name)
🚀 - make it an initializer on the `file` type

**Comment by mppf** [comment]
As I wrote in https://github.com/chapel-lang/chapel/issues/20143#issuecomment-1172738252  :

> In terms of naming, what bugs me about `openfd` (and `openfp`) is that they don't actually open anything at all. They just create a Chapel `file` that refers to a C or OS file. So I don't think they should have `open` in the name.

**Comment by lydia-duncan** [comment]
Brad, how do you feel about `_file` being a wrapper for a C type that is inserted by the compiler instead of defined somewhere in our modules, relative to moving it to CTypes or moving it to CTypes and renaming it?

**Comment by bradcray** [comment]
Would it be user facing/visible?  If not, I don't have a problem with that (though I'd probably call it `chpl_`-something to avoid naming conflicts).  If so, that doesn't seem like a very attractive/clear name for a user-facing identifier.

**Comment by lydia-duncan** [comment]
It's currently accessible in all programs today, though it isn't in our documentation except for places where it is used (which I think is just `openfp`'s argument).

Fwiw, I am strongly in favor of renaming it, I was just trying to tackle one issue at a time in the voting and the first issue imo is that it is a compiler-defined type when I'm not sure it needs to be.

**Comment by lydia-duncan** [comment]
I think they'd have to define an extern function that returns the type (and to do that, they'd need to know that the type exists)

**Comment by lydia-duncan** [comment]
But personally, it seems like a user is more likely to work with `FILE *` than a file descriptor, coming from C

**Comment by bradcray** [comment]
> But personally, it seems like a user is more likely to work with FILE * than a file descriptor, coming from C

I agree with that which is why I prefer `c_ptr(c_FILE)` to describe this type when it shows up in user routines.  But re-reading your comments, it sounds like that's your preference as well and then when you were asking how I felt about the compiler inserting `_file` and not documenting it, you weren't championing that approach, but looking for others to agree that we should move away from it (?).

**Comment by bradcray** [comment]
(note that a lot of this I/O code probably pre-dated support for `c_ptr` which seems like a likely rationale for the state of some of the types...)

**Comment by lydia-duncan** [comment]
> But re-reading your comments, it sounds like that's your preference as well and then when you were asking how I felt about the compiler inserting `_file` and not documenting it, you weren't championing that approach, but looking for others to agree that we should move away from it (?).

Yup, that's right

**Comment by lydia-duncan** [comment]
We discussed this issue in a group meeting today and were able to come to what I think are all the decisions necessary for this topic:

- Should we move the `_file` type to the modules, if possible?  No one voted against it, and 7 people were in favor
- What should the name of the `_file` type be?
  - 4 did not like `_file` at all
  - 2 liked `c_file`
  - 11 liked `c_FILE`
  - no one liked `C_FILE`
- What module should `_file` be moved to?
  - 1 person was against the IO module
  - 7 people liked putting it in the CTypes module
  - No one thought there was a better place
- How should `openfp` be renamed?
  - 3 people did not like `openFP`
  - 2 people did not like `openFp`
  - 1 person did not like `openFilePtr`
  - 1 person liked the factory function on the file type option but they also liked the initializer option
  - 11 people liked making it an initializer instead

With these decisions, I believe this change can be implemented

**Comment by jeremiah-corrado** [comment]
The `_file` type adjustments associated with this issue have been implemented in : #20775 

`openfp` still needs to be deprecated and replaced with an initializer on the `file` type.
